### PR TITLE
ORCA: Improve cardinality estimation for Dynamic Table Scan under SPE

### PR DESCRIPTION
**Pull request by l-wang**
ORCA: Improve cardinality estimation for Dynamic Table Scan under SPE

Previously, Orca always generates cardinality estimation for the
CLogicalDynamicGetBase operator by applying the query predicate stats
on top of the root table stats.  This makes sense for Dynamic Bitmap
Heap Scan and Dynamic Index Scan because the input rows for those
dynamic scans are filtered exactly by the query predicate.  However,
Dynamic Table Scan (CLogicalDynamicGet) is slightly different.
Although the query predicate helps partition pruning, we scan every
tuple of the selected partitions regardless the selectivity of the
query predicate within a leaf partition.  Therefore, for static
partition elimination, we should use the combined statistics of the
selected leaf partitions to estimate cardinality of
CLogicalDynamicGet.

To achieve the above goal there are two options:

Option 1: When deriving stats of CLogicalDynamicGet, use the selected
partitions' disjunctive constraints as the stats filter for the root
table.

Option 2: When deriving stats of CLogicalDynamicGet, retrieve and
combine stats of each selected leaf partition.

This patch went with option 1. The disjunctive constraints can be
easily recorded as part of static partition elimination.  The downside
of option 2 is a) more I/Os of the MD accessor to retrieve leaf
partition statistics if a lot of leaf partitions are selected and b)
complicate code a little bit for column matching when retrieving and
merging statistics.

### CODE REVIEW

File: src/backend/gporca/data/dxl/minidump/PartTbl-SPE-DynamicTableScan-Range-Cost5.mdp [hunk 1]
```diff
@@ -0,0 +1,2313 @@
...
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+  <dxl:Thread Id="0">
+    <dxl:Comment><![CDATA[
+	Objective: We should derive statistics of a Dynamic Table Scan with SPE
+	using the disjunctive constraints of the selected child partitions. In this
+	case, all partitions are selected because the filter predicate is
+	unsupported, CPhysicalDynamicTableScan in physical plan should cost arround
+	2456 with all 1101000 rows.
+
+	create table default_part_only (a int, b int, c varchar(10000)) partition by range (b)(default partition d);
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/gporca/data/dxl/minidump/PartTbl-SPE-DynamicTableScan-Range-Cost5.mdp]: Fixed.

**Review by dgkimura — COMMENTED** [review]
Nice work!

I've read the commit message a few times and I have a some questions to make
sure I understand:

Is the crux of the issue that the selectivity of the predicate is based on the
full table and when SPE eliminates portions of the table then the selectivity 
is no longer accurate?  But this isn't true for dynamic bitmap heap scan and
dynamic index scan because they don't eliminate partitions?

Is there anything wrong with applying your approach to use the partitions'
disjunctive constraints for both cases?  Just trying to justify the fork in the
code.

Is there a potential accuracy difference between option 1 and 2?  And if so, is 
there a worst case?

Are partial indexes handled correctly?

**Comment by l-wang** [comment]
> I've read the commit message a few times and I have a some questions to make sure I understand:

OK the fact that you have to read the commit message a few times means the commit message is bad. I will rephrase the commit message and make it easier to read.

> Is the crux of the issue that the selectivity of the predicate is based on the full table and when SPE eliminates portions of the table then the selectivity is no longer accurate? But this isn't true for dynamic bitmap heap scan and dynamic index scan because they don't eliminate partitions?

Not exactly. I think It will be help demonstrate the crux of the issue if we look at the physical plans:

CPhysicalDynamicTableScan:
```
+--CPhysicalMotionGather(master)   rows:934   width:10042  rebinds:1   cost:473.388452   origin: [Grp:4, GrpExpr:4]
   +--CPhysicalFilter   rows:934   width:10042  rebinds:1   cost:438.529666   origin: [Grp:4, GrpExpr:3]
      |--CPhysicalDynamicTableScan "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Scan Id: 1   rows:934   width:10042  rebinds:1   cost:432.720016   origin: [Grp:0, GrpExpr:1]
      +--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
         |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
         +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
```

CPhysicalDynamicBitmapTableScan:
```
+--CPhysicalMotionGather(master)   rows:934   width:10042  rebinds:1   cost:507.287870   origin: [Grp:4, GrpExpr:4]
   +--CPhysicalFilter   rows:934   width:10042  rebinds:1   cost:472.429084   origin: [Grp:4, GrpExpr:2]
      |--CPhysicalDynamicBitmapTableScan "foo" ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Scan Id: 1   rows:934   width:10042  rebinds:1   cost:466.619434   origin: [Grp:6, GrpExpr:1]
      |  |--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
      |  |  |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
      |  |  +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
      |  +--CScalarBitmapIndexProbe   Bitmap Index Name: ("foo_idx_bitmap")   origin: [Grp:5, GrpExpr:0]
      |     +--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
      |        |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
      |        +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
      +--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
         |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
         +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
```

CPhysicalDynamicIndexScan:
```
+--CPhysicalMotionGather(master)   rows:934   width:10042  rebinds:1   cost:58.165746   origin: [Grp:4, GrpExpr:6]
   +--CPhysicalFilter   rows:934   width:10042  rebinds:1   cost:23.306960   origin: [Grp:4, GrpExpr:3]
      |--CPhysicalDynamicIndexScan   Index: ("foo_idx_btree"), Table: ("foo"), Columns: ["a" (0), "b" (1), "c" (2), "ctid" (3), "xmin" (4), "cmin" (5), "xmax" (6), "cmax" (7), "tableoid" (8), "gp_segment_id" (9)] Scan Id: 1   rows:934   width:10042  rebinds:1   cost:17.497310   origin: [Grp:7, GrpExpr:1]
      |  +--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
      |     |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
      |     +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
      +--CScalarCmp (=)   origin: [Grp:3, GrpExpr:0]
         |--CScalarIdent "b" (1)   origin: [Grp:1, GrpExpr:0]
         +--CScalarConst (2)   origin: [Grp:2, GrpExpr:0]
```

If you compare the CPhysicalDynamic{Table|BitmapTable|Index}Scan operators, you may notice that both `CPhysicalDynamicBitmapTableScan` and `CPhysicalDynamicIndexScan` has `CScalarCmp` for `b = 2` as their children. That makes sense because that predicate helps reducing the output tuples of those two types of scans. However, since this is post SPE, `CPhysicalDynamicTableScan` is meant to be outputing every tuple of the selected leaf partitions, therefore no predicate shall be considered inside the Dynamic Table Scan. The physical plan shape of the Dynamic Table Scan was correct, however, when we derive stats for its CLogical ancestor `CLogicalDynamicGet`, we nevertheless applied the predicate, and that causes under-estimation for the cardinality and cost for `CPhysicalDynamicTableScan`. In this specific example, `CPhysicalDynamicTableScan` estimated `934` rows, while we should expect ~1M rows. Because the cardinality for this node is greatly under estimated, the corresponding cost is also under estimated. In this example, the overall estimated row count is correct because applying a supported predicate is idempotent, but the overall cost is incorrect.

The setup and queries for the above plans for reference:
```sql
drop table foo;
create table foo (a int, b int, c varchar(1000000)) partition by range (b) (start (1) end (4) every (2));
create index foo_idx_bitmap on foo using bitmap(b);
insert into foo select i, 1, repeat('a',10000) from generate_series(1, 1000000)i;
insert into foo select i, 2, repeat('a',10000) from generate_series(1, 1000)i;
analyze foo;

set optimizer_print_plan to on;
set client_min_messages to 'log';
explain select * from foo where b = 2;
set optimizer_enable_dynamictablescan to off;
explain select * from foo where b = 2;

create index foo_idx_btree on foo using btree(b);
explain select * from foo where b = 2;

reset optimizer_print_plan;
reset client_min_messages;
reset optimizer_enable_dynamictablescan;
```

> Is there anything wrong with applying your approach to use the partitions' disjunctive constraints for both cases? Just trying to justify the fork in the code.

I hope the above example could also answer this question. Let me know if not.

> Is there a potential accuracy difference between option 1 and 2? And if so, is there a worst case?

A potential accuracy concern I had in mind for option 1 was the case where the CExpression of the disjunctive constraints of leaf partitions is unsupported. However, I couldn't come up with such expressions for either List or Range partition bounds. 

> Are partial indexes handled correctly?

Good thought! But this is GPDB7 we got rid of partial indexes.

File: src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDynamicGet.h [hunk 2]
```diff
@@ -32,6 +32,13 @@ class CColRefSet;
 //---------------------------------------------------------------------------
 class CLogicalDynamicGet : public CLogicalDynamicGetBase
 {
+protected:
+	// Disjunction of selected child partition's constraints after static pruning
+	CConstraint *m_partition_cnstrs_disj = nullptr;
+
+	// Has done static pruning
+	BOOL m_static_pruned = false;
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/gporca/libgpopt/include/gpopt/operators/CLogicalDynamicGet.h:40]: Thanks for pointing this out! Will make the change!

File: src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGet.cpp [hunk 3]
```diff
@@ -50,14 +50,15 @@ CLogicalDynamicGet::CLogicalDynamicGet(CMemoryPool *mp)
...
+	ULONG ulPartIndex, CColRefArray *pdrgpcrOutput,
+	CColRef2dArray *pdrgpdrgpcrPart, IMdIdArray *partition_mdids,
+	CConstraint *partition_cnstrs_disj, BOOL static_pruned = false)
 	: CLogicalDynamicGetBase(mp, pnameAlias, ptabdesc, ulPartIndex,
-							 pdrgpcrOutput, pdrgpdrgpcrPart, partition_mdids)
+							 pdrgpcrOutput, pdrgpdrgpcrPart, partition_mdids),
+	  m_partition_cnstrs_disj(partition_cnstrs_disj),
+	  m_static_pruned(static_pruned)
 {
 }
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGet.cpp:63]: You are right! `static_pruned` seems unnecessary will remove it.

**Review by dgkimura — COMMENTED** [review]
Thanks for your explanations.  It makes more sense to me now.

Does this approach need to be updated when we re-enable support for multi-level partition tables?  And if so, what might that change entail?

File: src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGetBase.cpp [hunk 4]
```diff
@@ -265,6 +266,29 @@ CLogicalDynamicGetBase::PstatsDeriveFilter(CMemoryPool *mp,
 		return pstatsFullTable;
 	}
 
+	if (COperator::EopLogicalDynamicGet == exprhdl.Pop()->Eopid())
```
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGetBase.cpp]: done.

**Comment by l-wang** [comment]
> Does this approach need to be updated when we re-enable support for multi-level partition tables? And if so, what might that change entail?

@dgkimura No we don't need to change this logic when we re-enable support for multi-level partition tables. For multi-level partitioning, after static partition pruning we will still get a list of selected leaf partitions.

File: src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGet.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (l-wang) [reply src/backend/gporca/libgpopt/src/operators/CLogicalDynamicGet.cpp]: My comment above was wrong. I got it right the first time, the `static_pruned` flag is necessary. We could have a null `partition_cnstrs_disj` but a true `static_pruned` if the default partition is the only partition. `PartTbl-SPE-DynamicTableScan-Range-Cost4.mdp` was supposed to be that test case but I copy and pasted the wrong file. I fixed that and added the assertion you suggested. Your suggestion was correct! Thank you! I dropped the commits I made yesterday and force pushed. Sorry about that!

**Review by dgkimura — APPROVED** [review]
Commit 11e2525 (last refactor) is causing MPD test failures.  Otherwise, looks good to me.  And thanks again for the detailed answers!

**Pull request merged**